### ISSUE TITLE
Data centering for TruncatedSVD?

### ISSUE THREAD

**Issue by gergo-szabo** [YingfanWang/PaCMAP#140]
The centering kills the TruncatedSVD's main "feature".

https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.TruncatedSVD.html
"Contrary to PCA, this estimator does not center the data before computing the singular value decomposition. This means it can work with sparse matrices efficiently."

Code:
https://github.com/YingfanWang/PaCMAP/blob/d20a403473f756970094534f2464ad95ce1902f7/source/pacmap/pacmap.py#L376

**Comment by gergo-szabo** [comment]
Also centering before PCA?
https://github.com/YingfanWang/PaCMAP/blob/d20a403473f756970094534f2464ad95ce1902f7/source/pacmap/pacmap.py#L388

Sklearn PCA does it by default.
https://github.com/scikit-learn/scikit-learn/blob/d3898d9d57aeb1e960d266613a2e31b07bca39d7/sklearn/decomposition/_pca.py#L570

**Comment by williamsyy** [comment]
@gergo-szabo Thanks for your interest in PaCMAP. I get your point that it might be a bit redundant here to center the dataset if truncated SVD is applied. Is there a specific usage that you have dealt with might be influenced by current setup?

**Comment by gergo-szabo** [comment]
No specific usecase at the moment which would be influenced by this issue. I was just checking your PCA implementation for my text embeddings.

preprocess_X function has two branch:

- One for high dim embeddings (> 100) or Hamming distance (often parse)
```
Centers data
Uses sklearn TruncatedSVD
```
Sklearn TruncatedSVD is mathematically equal to sklearn PCA with centering. (Implementation-wise is different. eg.: solver type).
TruncatedSVD gives better results if data is sparse (eg.: "bag of words" type of data) but this repo basically uses PCA.


- Else branch, all other option
```
Centers data
Scales it to -1..1
Uses sklearn PCA
```
Since you are scaling to -1..1 the centering is ok here if scaling has a meaning. Sklearn PCA does a redundant centering but that is a rounding issue maximum.


If you check UMAP's implementation they skip centering (and scaling) for both branch:
https://github.com/lmcinnes/umap/blob/0ff1d1e79904935b8711f7c60d477b21e8f0bc49/umap/umap_.py#L1099-L1104
